### ISSUE TITLE
False Positive of `clippy::missing_const_for_fn`

### ISSUE THREAD

**Issue by yhx-12243** [rust-lang/rust-clippy#14013]
### Summary

The following function could not be const now:
```rust
pub fn get(x: &Arc<()>) -> &() {
    x
}
```

### Lint Name

missing_const_for_fn

### Reproducer

I tried this code:

```rust
#![warn(clippy::missing_const_for_fn)]

use std::sync::Arc;

pub fn get(x: &Arc<()>) -> &() {
    x
}

fn main() {}
```

I saw this happen:

```
warning: this could be a `const fn`
 --> 1.rs:5:1
  |
5 | / pub fn get(x: &Arc<()>) -> &() {
6 | |     x
7 | | }
  | |_^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#missing_const_for_fn
note: the lint level is defined here
 --> 1.rs:1:9
  |
1 | #![warn(clippy::missing_const_for_fn)]
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: make the function `const`
  |
5 | pub const fn get(x: &Arc<()>) -> &() {
  |     +++++

warning: 1 warning emitted
```

I expected to see this happen:

\<nothing\> (because adding `const` will cause a compilation error)

### Version

```text
rustc 1.86.0-nightly (99db2737c 2025-01-16)
binary: rustc
commit-hash: 99db2737c91d1e4b36b2ffc17dcda5878bcae625
commit-date: 2025-01-16
host: aarch64-apple-darwin
release: 1.86.0-nightly
LLVM version: 19.1.7
```

**Comment by alex-semenyuk** [comment]
Suppose can be considered as duplication #14020
